### ISSUE TITLE
[DEPR]: karma-selenium-webdriver-launcher

### ISSUE THREAD

**Issue by brian-smith-tcril** [openedx/public-engineering#248]
### RFC Start Date

2023-03-26

### Target Plan Accepted Date

2023-04-09

### Target Transition Unblocked Date

2023-04-09

### Earliest Breaking Changes Unblocked Date

2023-04-09

### Earliest Open edX Named Release Without This Functionality

Ulmo - 2025-06 

### Rationale

In the process of upgrading `edx-platform` to node 18, `npm install` failed because the `selenium-webdriver` peer dependency in `karma-selenium-webdriver-launcher` is set to  `^2.44.0`, but we have `3.4.0` as a dev dependency.

To work around this issue, we forked `karma-selenium-webdriver-launcher` to the `openedx` GH org, and updated the peer dependency to allow newer versions of `selenium-webdriver` (https://github.com/openedx/karma-selenium-webdriver-launcher/pull/1/files)

This package is unmaintained (hasn't been updated in 8 years), and there are other ways to handle this problem.

### Removal

* The https://github.com/openedx/karma-selenium-webdriver-launcher repo
* All usage of `karma-selenium-webdriver-launcher` in `edx-platform` and any other repos using it

### Replacement

TBD

**Comment by tonybusa** [comment]
assign me

**Comment by tonybusa** [comment]
I've started on this ticket. It appears the only references for the package are within edx-platform (other than the library itself):
https://github.com/search?q=org%3Aopenedx+karma-selenium-webdriver-launcher&type=code 

After removing the references, I'm able to build and run the platform. But I assume we rely on this package for running tests? I'll continue my research within this article: https://docs.openedx.org/projects/edx-platform/en/latest/concepts/testing/testing.html#running-javascript-unit-tests 

The ticket mentions "there are other ways to handle this problem". @brian-smith-tcril  if you have insight on this please let me know. Since the replacement says "TBD" I'll continue to tackle the testing problem, if that's is the correct verification of this ticket being "done", and will report back any findings in another comment. Thanks!

**Comment by tonybusa** [comment]
I was able to get the karma tests running locally from edx-platform with the following commands (I use Chrome as my browser):
```
karma start  cms/static/karma_cms.conf.js --browsers=Chrome --single-run=false
```
and
```
 karma start lms/static/karma_lms.conf.js --browsers=Chrome --single-run=false 
```
The LMS command fails unless I comment out the following spec files that are using Webpack:

from file: ```lms/static/karma_lms.conf.js```
```
  // Define the Webpack-built spec files first
   {pattern: 'course_experience/js/**/*_spec.js', webpack: true},
   {pattern: 'js/learner_dashboard/**/*_spec.js', webpack: true},
   {pattern: 'js/student_account/components/**/*_spec.js', webpack: true},
   {pattern: 'completion/js/**/*_spec.js', webpack: true},
```
Also mentioning that this file can be removed from the sourceFiles array in the same config file since it doesn't reference any tests:
```
{ pattern: "learner_profile/**/!(*spec).js" },
```
I suspect that may be the basis of this issue, since the ```test-karma-webpack``` command gets ignored in the CI anyways. Perhaps these are unrelated but figured I'd mention it.

Noting the results from the LMS run: 781 specs, 8 failures, 6 pending specs
**Failures:**
1) discovery.views.CourseCard renders
2-8)  ![Image](https://github.com/user-attachments/assets/f6af1822-db2f-417e-a231-9dd4ecbb9135)

**Comment by tonybusa** [comment]
Posting the link the PR I created and some notes from @brian-smith-tcril.

https://github.com/openedx/edx-platform/pull/36589

Brian indicated that https://github.com/openedx/edx-platform/blob/2eb0adb4402b4ca30a6444a0a9c1b15646924923/package.json#L27 as in ```test-lms``` doesn't get run in the CI. I'm curious if we want coverage on the karma tests that don't require webpack as noted in the comment above. However, there would need to be fixes implemented for the 8 tests failing above. Waiting to see how my PR does in the CI before following up in the working group next week!

**Comment by robrap** [comment]
[inform] This DEPR is in the Draft state. It can't move forward without a DEPR coordinator to help update it to match the updated DEPR template. No action is required, unless anyone wants to move this out of Draft.

**Comment by e0d** [comment]
Hi @robrap I think we need to prioritize getting issues that are older and have effort put into them over the line.  What's the best way to do that?  I'm happy to steward this issue in the case, but I want to make sure we're not creating a pile of issues that need similar interventions when we'd all be better off getting them to done.

**Comment by robrap** [comment]
@e0d: Getting a DEPR WG coordinator assigned to the ticket is the best way, and @kdmccormick and @feanil have been seeing through most of the axim-initatied tickets. Maybe one of them can coordinate and get this ticket into the right place. The updated process doesn't really add new work, but just makes the transitions and operator impact more clear. Thanks!

**Comment by robrap** [comment]
Also, @e0d, just so you are clear - the interventions are minor - and will just help 2U and MIT be clear how and when we will be impacted, and how to prepare properly. There are way too many tickets in an unclear state to know at this point. If this ticket moves along, at least it will be on our radar.

**Comment by feanil** [comment]
I've updated this ticket, I've moved it to breaking changes unblocked.  I think this might be a good example of a maintenance ticket we didn't have another good way of tracking historically.  This is re-working tests and test asset management and does not impact the end product necessarily.  Unless we find otherwise as we proceed with the removal I think this is good to go.  If we find that we need to make changes that might break the app in some way, please chime back in here and we can figure out if we need to announce that before proceeding.  But for now this can move forward.

@tonybusa @brian-smith-tcril do you have everything you need to move forward with this? It seems like maybe there is a question about how to proceed with running some of the older tests that have been disabled? Let me know if you already have enough info to make this decision yourselves or if you need further input/guidance.

**Comment by brian-smith-tcril** [comment]
I think for this DEPR

In scope:
* Find `edx-platform` tests that are currently running in CI that rely on `karma-selenium-webdriver-launcher`
* Update those tests to manually do what `karma-selenium-webdriver-launcher` does (based on the code in https://github.com/openedx/karma-selenium-webdriver-launcher/blob/master/index.js)

Out of scope:
* Anything related to tests that are not currently running

Edit: based on https://github.com/openedx/edx-platform/pull/36589 and https://github.com/openedx/edx-platform/issues/35956 it seems as though *no* tests that are currently running for `edx-platform` rely on `karma-selenium-webdriver-launcher`. I'm happy to move forward with removing the dependency and archiving the `karma-selenium-webdriver-launcher` repo, but it may add some complexity to https://github.com/openedx/edx-platform/issues/35956. If https://github.com/openedx/edx-platform/pull/36589 lands without any test changes it would be good to make a note on https://github.com/openedx/edx-platform/issues/35956 explaining that some of the disabled tests likely depend on `karma-selenium-webdriver-launcher`.

**Comment by robrap** [comment]
@brian-smith-tcril: [tip] Github will expand PR/issue names in github links that are included in bullets, but uses a shorthand if bullets are not used. I'm adding your links using bullets as a reference for your comment.
- https://github.com/openedx/edx-platform/pull/36589
- https://github.com/openedx/edx-platform/issues/35956

@brian-smith-tcril @kdmccormick: I wonder with or without this DEPR if we need some sort of catalog of the actual missing test coverage and prioritization of which tests matter or not? That question seems both related and independent of this DEPR.

**Comment by brian-smith-tcril** [comment]
> [tip] Github will expand PR/issue names in github links that are included in bullets, but uses a shorthand if bullets are not used. I'm adding your links using bullets as a reference for your comment.

Yeah, I just wanted to reference them inline and I know you can hover to see details

![Image](https://github.com/user-attachments/assets/96af7232-899c-48c9-9acf-02aa80050e66)

_____

> if we need some sort of catalog of the actual missing test coverage and prioritization of which tests matter or not?

That feels like it could be a part of https://github.com/openedx/edx-platform/issues/35956 (as opposed to a part of this DEPR).

**Comment by robrap** [comment]
@brian-smith-tcril:

> Yeah, I just wanted to reference them inline and I know you can hover to see details

I wasn't aware of hovering. I guess I click too fast? Thanks.

> That feels like it could be a part of https://github.com/openedx/edx-platform/issues/35956 (as opposed to a part of this DEPR).

Agreed.

**Comment by feanil** [comment]
I think we should put a note in https://github.com/openedx/edx-platform/issues/35956 to indicate that we have removed this particular test dependence since those tests aren't running and that it shouldn't be added back?  Once that comment is made, the dependency can be removed from package.json?

**Comment by robrap** [comment]
@feanil: For the flaky test process, we got in the habit of deleting tests that weren't running so the code base wasn't filled with skipped tests that gave a false perception of coverage. Would it make sense to remove all tests that aren't running as part of the DEPR, and have https://github.com/openedx/edx-platform/issues/35956 be "fix and restore", rather than "fix and re-enable"?

**Comment by feanil** [comment]
I think that will be just extra churn right now because they're about to actively work on it so I don't think that's necessary.  We can re-visit in a few weeks but @tonybusa it may be worth talking to @ttqureshi who will also be trying to get some of these tests running again.

**Comment by ttqureshi** [comment]
I took a time-boxed look into this issue [couple of weeks ago](https://github.com/openedx/edx-platform/issues/35956#issuecomment-2969248222) in an attempt to get `npm run test-xmodule-webpack` command working and see what's breaking, but unfortunately couldn't figure it out, possibly because I don’t have a broader JS/Webpack background or detailed familiarity with the `karma-webpack` internals.

**Comment by dianakhuang** [comment]
Has any progress been made on this or has it been deprioritized?

**Comment by feanil** [comment]
This work is done.  The plugin has been removed
